Fix - Empty workspace chats appear in the LHN for the admin

FitseTLT opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/58859
PROPOSAL: https://github.com/Expensify/App/issues/58859#issuecomment-2741639565


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
Precondition: use most recent mode.

1. Create a workspace
2. Invite a member to the workspace
3. Go to the Inbox
4. While on another chat verify that all empty own and employee workspace chats are not available in the LHN

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
Same as above
### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
Same as above
- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I added [unit tests](https://github.com/Expensify/App/blob/main/tests/README.md) for any new feature or bug fix in this PR to help automatically prevent regressions in this user flow.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/8cf8bc87-dca3-49b5-9294-b788a081a2b4


</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/a9a17fe8-a58b-40da-a394-f29b36cfce8a


</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/6ad41d98-b366-43f6-8532-8690cee6e8c2


</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/edb7f45f-3595-4864-afad-f567854cadd7


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/a6d92b4b-4a31-4515-a3ae-3ecec493b76f



</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->


https://github.com/user-attachments/assets/57795e7d-9ed2-4cc4-9086-fb4be4f797bf


</details>

FitseTLT committed:
avoid empty policy chats on LHN

FitseTLT committed:
added test

FitseTLT commented:
@trjExpensify Do u agree with the above? Just to make sure.

trjExpensify commented:
Oh, that's weird. How's that actually happening? 

1. With this change, empty workspace chats don't show in the LHN. 
2. You delete a workspace || remove a member. 
3. The archived workspace chat(s) then somehow appear in the LHN?

trjExpensify commented:
Yeah, but how or why is that happening to make them magically appear when they weren't there in the first place with this change: 

> 1. With this change, empty workspace chats don't show in the LHN.
> 2. You delete a workspace || remove a member.
> 3. The archived workspace chat(s) then somehow appear in the LHN?

trjExpensify commented:
Yeah, I'm referring to this PR and executing these steps:

1. An admin creates a workspace > invites a member 
2. The empty workspace chat doesn't appear in the LHN with this PR change.
3. The admin then removes the member from the workspace 
4. **Why does this empty archived workspace chat suddenly appear in the LHN now?**

FitseTLT commented:
@trjExpensify I was just asking for expected result to make the changes. And from the above comment I think you are confirming that you don't want empty archived chat to appear I will make the necessary changes assuming that. Please correct me if that is not the case Thx 👍

FitseTLT commented:
@trjExpensify @parasharrajat The root cause is the code that handles the archived reports comes before the code that excludes empty chats. 
https://github.com/Expensify/App/blob/db8867818cbe54da8c83dbae535d04c66fbf12af/src/libs/ReportUtils.ts#L7470-L7477
So I was thinking we should change the order assuming any type of empty chats that we are excluding from the list shouldn't be included into the list because it is archived. Can u think of any case (report type) we wouldn't want this change to be applied? Otherwise, I can make the change only for policy chats to be safe 🤔

FitseTLT committed:
apply excludeEmptyChats check before is archived check

FitseTLT commented:
Yep applied

FitseTLT committed:
update test

FitseTLT commented:
https://github.com/user-attachments/assets/60e23277-9587-492a-a0c3-b4dc0affc207

Please check again.

FitseTLT commented:
> Also, the chat should not be hidden from the member whose workspace chat it is.

This is not clear. Are u pointing own workspace chats?

FitseTLT committed:
Merge branch 'main' into fix-avoid-empty-workspace-chats-on-LHN

FitseTLT commented:
> No, let say I am member of a workspace and I have workspace chats. Those should be visible to me.

Why will we make an exception if it is an empty workspace chat? cc @trjExpensify

FitseTLT commented:
> Looks like pR does not hide the member expense chats which are empty **For old members**. This is a old member, even after I removed and added back it visible.

> ![image](https://private-user-images.githubusercontent.com/24370807/429166606-d3965b12-c31e-425d-908a-595dd3f89e97.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM1MzUzNjEsIm5iZiI6MTc0MzUzNTA2MSwicGF0aCI6Ii8yNDM3MDgwNy80MjkxNjY2MDYtZDM5NjViMTItYzMxZS00MjVkLTkwOGEtNTk1ZGQzZjg5ZTk3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDAxVDE5MTc0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY0OTFiNGIxMWQzNDdiZjc1YWMxZWZkZjJmMGMxNmJiZWYzNmU5OGU4ZGY0NjllOWRmNDUzYjgyNmNiYzFiNjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MZQlxIbm4k8OuybZE3kT-gNmvhBnxK3KAxGDfGhJBFg)

@parasharrajat  Yes u are right but this happens because for an old member the lastMessageText of the report is not empty `"lastMessageText": "test.engfitsum@gmail.com closed this report",` . This is because there is a `CLOSED` report action that is hidden. Unfortunately this a limitation of our App. Ideally you could determine whether a report is empty by checking the report actions list but report actions of a report are not necessarily available in onyx, for instance, on sign in before you open a report.

> 
> Did you check whether chats where messages were deleted are getting hidden as well or not?
> 

As long as the BE correctly resets the lastMessageText it works perfectly.

trjExpensify commented:
>No, let say I am member of a workspace and I have workspace chats. Those should be visible to me.

If the workspace chat is empty? No, don't show it in the LHN. Chances are if you're an invited member it's not, because the workspace welcomeNote is posted in it on being invited.

FitseTLT commented:
The above answer already aligns with the code. Yes

rlinoz commented:
@FitseTLT can you merge main to fix the conflicts, please?

@parasharrajat can you complete the review, please?

FitseTLT committed:
Merge branch 'main' into fix-avoid-empty-workspace-chats-on-LHN

FitseTLT commented:
Done

FitseTLT committed:
resolve conflict

trjExpensify commented:
@parasharrajat over to you, let's get this done.

FitseTLT commented:
[Read](https://github.com/Expensify/App/pull/59304#issuecomment-2770629723)

rlinoz commented:
started a thread here https://expensify.slack.com/archives/C01GTK53T8Q/p1744900572184299 to talk about the pr more quickly.

trjExpensify commented:
Can you confirm what we're looking at here?

trjExpensify commented:
Thanks, got it. Why does the chat disappear when you click off it, and you're in most recent mode right? That's not expected.

FitseTLT commented:
> Thanks, got it. Why does the chat disappear when you click off it, and you're in most recent mode right? That's not expected.

@trjExpensify Why will it not disappear? Isn't it expected for empty workspace chat?

trjExpensify commented:
It's not empty for the workspace member, the welcomeNote is there as a message whispered to them: 

<img width="1511" alt="image" src="https://github.com/user-attachments/assets/28bcce6c-58c2-4f31-a0cf-9d8a63a6cbfd" />

FitseTLT committed:
Merge branch 'main' into fix-avoid-empty-workspace-chats-on-LHN

FitseTLT commented:
> It's not empty for the workspace member, the welcomeNote is there as a message whispered to them:
> 
> <img alt="image" width="1511" src="https://private-user-images.githubusercontent.com/16232057/436565615-28bcce6c-58c2-4f31-a0cf-9d8a63a6cbfd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU0MTYzNzMsIm5iZiI6MTc0NTQxNjA3MywicGF0aCI6Ii8xNjIzMjA1Ny80MzY1NjU2MTUtMjhiY2NlNmMtNThjMi00ZjMxLWEwY2YtOWQ4YTYzYTZjYmZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIzVDEzNDc1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE1ZTk5ZTAwZWYxMzlhNGNkZTJlNzg4YjVlN2RkZWZjN2JiMzM2MjNhN2MwOTRmN2ZlOTYzNjkzMjE3MTEyNDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Cwwr8RmB3juZE3MilC1Zx77DiIe28IFa8y5gEEuUdsw">

But we don't consider whisper actions as visible last actions 

https://github.com/Expensify/App/blob/d59ec4c8eff6b92ee917dfd5f2afc56b1455a824/src/libs/ReportActionsUtils.ts#L903

Note `only visible to you` text.

Are u suggesting we change that logic here?

rlinoz commented:
@trjExpensify https://github.com/Expensify/App/pull/59304#issuecomment-2824968996

Also, we have this thread if you wanna use it for a quicker discussion flow https://expensify.slack.com/archives/C01GTK53T8Q/p1744900572184299

trjExpensify commented:
@rodrigo, remind me. What’s the latest expected behaviour for whispered messages? Namely:

- Should the whisper message appear in the LHN preview for the person it was whisperedTo? 
- Should the chat be bold and unread for the person it was whisperedTo? 

I think whispers have been through the ringer a few times with different performance considerations etc, so I’m not 100% sure what the latest expectation is.

trjExpensify commented:
>I am not sure why though

Yeah, was kinda' hoping for insight on that. I can find this which kinda' suggests we notify: https://github.com/Expensify/Web-Expensify/blob/b2165c7f4d4835eee8396171e150a46a526485cd/lib/ReportAPI.php#L1173-L1186

Anyhow, overall here, I think the ideal expectation for whisper messages is that they behave like comments, but only for the accountIDs the message was whisperedTo: 

- Chat is unread for the member the message is whisperedTo only
- LHN preview shows the whispered message to the whisperedTo participant only
- The chat therefore isn't treated "empty" and shows in the LHN for the whisperedTo participant only

@deetergp I don't suppose this is something on the radar with the whole "actionable system message" framework stuff?

deetergp commented:
@trjExpensify No, that's not on the radar currently. Isn't "Most recent" mode supposed to contain all of the chats you are currently a part of? Even if there's no activity in it? 🤔

trjExpensify commented:
> Isn't "Most recent" mode supposed to contain all of the chats you are currently a part of? Even if there's no activity in it? 🤔

Nah, empty and archived chats for example do not get included.

trjExpensify commented:
>Let's discuss this on Slack https://expensify.slack.com/archives/C01GTK53T8Q/p1744900572184299.

Replied. This PR might not be necessary anymore. Let's discuss there.

FitseTLT commented:
Closing as the changes are no longer necessary.
